src/storage/copier/csv_reader.cpp

```diff
@@ -0,0 +1,515 @@
+        return -1;
+    }
+handle_escape:
+    /* state: handle_escape */
+    // escape should be followed by a quote or another escape character
+    position++;
+    if (position >= bufferSize && !ReadBuffer(start, line_start)) {
+        errorMessage = StringUtils::string_format(
+            "Error in file {} on line {}: neither QUOTE nor ESCAPE is proceeded by ESCAPE.",
+            filePath, linenr);
+        return -1;
+    }
+    if (buffer[position] != csvReaderConfig->quoteChar &&
+        buffer[position] != csvReaderConfig->escapeChar) {
+        errorMessage = StringUtils::string_format(
+            "Error in file {} on line {}}: neither QUOTE nor ESCAPE is proceeded by ESCAPE.",
+            filePath, linenr);
+        return -1;
+    }
+    // escape was followed by quote or escape, go back to quoted state
+    goto in_quotes;
+carriage_return:
+    /* state: carriage_return */
+    // this stage optionally skips a newline (\n) character, which allows \r\n to be interpreted as
+    // a single line
+    if (buffer[position] == '\n') {
+        // newline after carriage return: skip
+        // increase position by 1 and move start to the new position
+        start = ++position;
+        if (position >= bufferSize && !ReadBuffer(start, line_start)) {
+            // file ends right after delimiter, go to final state
+            goto final_state;
+        }
+    }
+    if (finished_chunk) {
+        return columnSizes[0];
+    }
+    SkipEmptyLines();
+    start = position;
+    line_start = position;
+    if (position >= bufferSize && !ReadBuffer(start, line_start)) {
+        // file ends right after delimiter, go to final state
+        goto final_state;
+    }
+
+    goto value_start;
+final_state:
+    if (finished_chunk) {
+        return columnSizes[0];
+    }
+
+    if (column > 0 || position > start) {
+        // remaining values to be added to the chunk
+        AddValue(std::string(buffer.get() + start, position - start - offset), column,
+            escape_positions, has_quotes);
+        finished_chunk = AddRow(insertChunk, column, errorMessage);
+        SkipEmptyLines();
+        if (!errorMessage.empty()) {
+            return -1;
+        }
+    }
+
+    // final stage, only reached after parsing the file is finished
+    // flush the parsed chunk and finalize parsing
+    if (mode == ParserMode::PARSING) {
+        Flush(insertChunk);
+    }
+
+    end_of_file_reached = true;
+    return columnSizes[0];
+}
+
+uint64_t BufferedCSVReader::ParseCSV(common::DataChunk& insertChunk) {
+    std::string errorMessage;
+    auto numRowsRead = TryParseCSV(mode, insertChunk, errorMessage);
+    if (numRowsRead == -1) {
+        throw CopyException(errorMessage);
+    }
+    return numRowsRead;
+}
+
+uint64_t BufferedCSVReader::TryParseCSV(
+    ParserMode parserMode, DataChunk& insertChunk, std::string& errorMessage) {
+    mode = parserMode;
+    return TryParseSimpleCSV(insertChunk, errorMessage);
```

Comment:
The way `mode` is setting is confusing to me. Why pass `mode`, the class field, from `ParseCSV` into `TryParseCSV`, and then set it again inside `TryParseCSV`. Is `mode` supposed to change somewhere?